src/session.rs

```diff
@@ -443,12 +445,15 @@ impl Session {
                                     // with the same parameters for which they previously
                                     // failed. We want to clear the error in this case.
                                     self.error = None;
+
+                                    callback(PollNotification::FinishedSuccessfully);
                                 }
                                 Err(interpret_error) => {
-                                    log::error!(
-                                        "Interpreter failed with error: {}",
-                                        interpret_error,
-                                    );
+                                    let error_message = format!("{}", interpret_error);
+                                    callback(PollNotification::FinishedWithError(
+                                        error_message.clone(),
+                                    ));
+                                    log::error!("Interpreter failed with error: {}", error_message);
```

Comment:
Let's allocate less here. Currently this does 3 allocations (`format!`, `log::error!` and `.clone()`), but two will suffice:

```rust
log::error!(
    "Interpreter failed with error: {}",
    interpret_error,
);
callback(PollNotification::FinishedWithError(format!("{}", interpret_error)));
```